Sources/TensorFlow/Layers/Recurrent.swift

```diff
@@ -420,12 +426,146 @@ public struct RecurrentLayer<Cell: RecurrentLayerCell>: Layer {
   }
 }
 
+/// A type with values that support differentiable binary operations.
+///
+/// Used by `BidirectionalRecurrentLayer` as a generic requirement for merge functions.
+public protocol Mergeable: Differentiable, AdditiveArithmetic {
+  /// Adds two values and produces their sum.
+  ///
+  /// - Note: renaming `sum` to `+` results in a compiler crash when conforming `Tensor` to
+  /// `Mergeable` (SR-13229).
+  @differentiable
+  static func sum(_ lhs: Self, _ rhs: Self) -> Self
+}
+
+extension Tensor: Mergeable where Scalar: TensorFlowFloatingPoint {
+  /// Adds two values and produces their sum.
+  @differentiable
+  public static func sum(_ lhs: Tensor, _ rhs: Tensor) -> Tensor {
+    lhs + rhs
+  }
+}
+
+/// Adds two mergeables and produces their sum.
+@differentiable
+public func sum<T: Mergeable>(
+  _ first: T,
+  _ second: T
+) -> T {
+  T.sum(first, second)
+}
+
+public struct BidirectionalRecurrentLayer<Cell: RecurrentLayerCell>: Layer
+where Cell.TimeStepOutput: Mergeable {
+  public typealias Input = [Cell.TimeStepInput]
+  public typealias Output = [Cell.TimeStepOutput]
+  public typealias MergeFunction = @differentiable (Cell.TimeStepOutput, Cell.TimeStepOutput) -> Cell.TimeStepOutput
+
+  /// A wrapper around a `@differentiable` merge function.
+  ///
+  /// - Note: this exists as a workaround for runtime crashes regarding `@differentiable`function
+  ///   stored properties (TF-1122).
+  private class _MergeFunction {
+    var function: MergeFunction
+    init(_ function: @escaping MergeFunction) {
+      self.function = function
+    }
+  }
+
+  @noDerivative private let _mergeFunction: _MergeFunction
+  /// The forward recurrent layer.
+  public var forward: RecurrentLayer<Cell>
+  /// The backward recurrent layer.
+  public var backward: RecurrentLayer<Cell>
+  /// The differentiable function used for merging forward and backward recurrent layer outputs.
+  @noDerivative public var mergeFunction: MergeFunction {
+    _mergeFunction.function
+  }
+
+  /// Creates an instance from the given recurrent layer cell and merge function.
+  ///
+  /// - Note: The default merge mode in Keras is `concat`, which is not currently supported.
+  ///   When the `concat` mode is added, it can be used as the default value.
```

Comment:
Yeah, I will look into it and add tests too 👍

Reply:
Sorry for the delay, but I encouraged a weird bug. I have added `concatenate` merge function and tests for it, but at runtime I got this error:

```
Precondition failed: Axis must be in `-rank..<rank` when calling split(sizes:alongAxis:) (rank: 0, axis: 1): file swift-apis/Sources/TensorFlow/Operators/Basic.swift, line 142
2020-09-02 19:57:25.801775+0200 xctest[93460:1943115] Precondition failed: Axis must be in `-rank..<rank` when calling split(sizes:alongAxis:) (rank: 0, axis: 1): file swift-apis/Sources/TensorFlow/Operators/Basic.swift, line 142
```

It is caused by using `lhs.concatenated(with: rhs, alongAxis: 1)` instead of `lhs + rhs`, but I did not find the error yet.

Reply:
Is this issue fixed now that concatenation happens along axis `-1` instead of axis `1`?

Reply:
Unfortunately not. I was trying to create some minimal example that would reproduce the error but without any luck. It is weird because if I just replace `lhs.concatenated(with: rhs, alongAxis: X)` with `lhs + rhs`, it just works. And all tests for sum merge mode against Keras are passing. And I did not find any special code for concatenated method in Keras.

Reply:
I see, thanks for the prompt response.

A minimal example would be really helpful. When you call `lhs.concatenated(with: rhs, alongAxis: -1)` and it triggers the precondition failure, what are the shapes of `lhs` and `rhs`?

Reply:
I found out that it fails at `_vjpConcatenated`. I added the following prints into the function in Basic.swift:

```
@inlinable
  @derivative(of: concatenated)
  func _vjpConcatenated(
    with other: Tensor,
    alongAxis axis: Int
  ) -> (value: Tensor, pullback: (Tensor) -> (Tensor, Tensor)) {
    let posAxis = axis < 0 ? axis + rank : axis
    let splits = [shape[posAxis], other.shape[posAxis]]
    return (
      concatenated(with: other, alongAxis: axis),
      { result in
        print("self", self)
        print("other", other)
        print("result", result)
        print("posAxis", posAxis)
        print("splits", splits)
        let gradients = result.split(sizes: splits, alongAxis: axis)
        print("gradients", gradients)
        return (gradients[0], gradients[1])
      }
    )
  }
```

the output after running `testBidirectionalBasicRNNConcatenateMerge` is:

```
self [[0.81258523, -0.5553411, 0.27422628, 0.73577887]]
other [[-0.39687672, -0.94476587, -0.47644293, -0.82400316]]
result [[1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0, 1.0]]
posAxis 1
splits [4, 4]
gradients [[[1.0, 1.0, 1.0, 1.0]], [[1.0, 1.0, 1.0, 1.0]]]
self [[-0.19038992,  0.44098023, 0.062064234,  -0.9267066]]
other [[  -0.5327341, 0.0063607423,   -0.5311964,    0.2796056]]
result 0.0
posAxis 1
splits [4, 4]
Precondition failed: Axis must be in `-rank..<rank` when calling split(sizes:alongAxis:) (rank: 0, axis: -1): file /Users/peter.jung/Projects/swift-apis/Sources/TensorFlow/Operators/Basic.swift, line 142
2020-09-09 20:08:04.530498+0200 xctest[69563:4338687] Precondition failed: Axis must be in `-rank..<rank` when calling split(sizes:alongAxis:) (rank: 0, axis: -1): file /Users/peter.jung/Projects/swift-apis/Sources/TensorFlow/Operators/Basic.swift, line 142
```

So, for some reason, at some point, `result` is `0` and then `result.split(...)` fails.

Reply:
@dan-zheng I have finally some free time and I was able to create a script to reproduce the error along with equivalent (working) python script. Please look at https://github.com/tensorflow/swift-apis/issues/1087.

Reply:
@dan-zheng Thanks to the @xihui-wu, I implemented a workaround. Concat-merge mode is now fully functional and set as a default merge function. Please check out the PR, everything seems to be working now :) :tada: